src/main/java/com/google/sps/model/AuthenticatedHttpServlet.java

```diff
@@ -0,0 +1,107 @@
+  // Error message if user is not authenticated
+  protected static final String ERROR_403 = "Authentication tokens not present / invalid";
+
+  /**
+   * Verifies user credentials on GET (sending a 403 error in the case that the user is not properly
+   * authenticated).
+   *
+   * @param request Http request sent from client
+   * @param response Http response to be sent back to the client
+   * @throws IOException if there is an issue processing the request
+   */
+  @Override
+  public final void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    Credential googleCredential = loadCredential(request, response);
+    if (googleCredential != null) {
+      doGet(request, response, googleCredential);
+    }
+  }
+
+  /**
+   * Verifies user credentials on POST (sending a 403 error in the case that the user is not
+   * properly authenticated).
+   *
+   * @param request Http request sent from client
+   * @param response Http response to be sent back to the client
+   * @throws IOException if there is an issue processing the request
+   */
+  @Override
+  public final void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    Credential googleCredential = loadCredential(request, response);
+    if (googleCredential != null) {
+      doPost(request, response, googleCredential);
+    }
+  }
+
+  /**
+   * Handle GET request
+   *
+   * @param request HTTP request from client
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException is there is an issue processing the request
+   */
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    response.sendError(400, "GET is not supported");
+  }
+
+  /**
+   * Handle POST request
+   *
+   * @param request HTTP request from client
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException is there is an issue processing the request
+   */
+  public void doPost(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    response.sendError(400, "POST is not supported");
+  }
+
+  /**
+   * Get credential, or return 403 error if the credential is invalid
+   *
+   * @param request HttpRequest from client
+   * @param response Response to send to client
+   * @throws IOException if an issue occurs processing the response
+   */
+  private Credential loadCredential(HttpServletRequest request, HttpServletResponse response)
+      throws IOException {
+    Credential googleCredential = AuthenticationUtility.getGoogleCredential(request);
+    if (googleCredential == null) {
+      response.sendError(403, ERROR_403);
+    }
```

Comment:
It's a little weird for a method called `loadCredentials` to also send a response in some cases. I would prefer to just inline this logic in each of `doGet` and `doPost`, with an early exit if credentials are `null`.

I think a small amount of code duplication in (what is effectively) a utility class is OK.

Reply:
I agree - this might be a case of over-abstraction. I'm going to get rid of the helper.